File: java/UnionFind.java
import java.utils.*;

class UnionFind {
  private int[] parents, sizes;
  public UnionFind(int n) {
    parents = new int[n];
    for (int i = 0; i < n; i++) {
      parents[i] = i;
    }
    for (int i = 0; i < n; i++) {
      sizes[i] = 1;
    }
  }

  public int find(int a) {
    int cur = a;
    while (cur != parents[cur]) {
      cur = parents[cur];
    }
    int root = cur;
    cur = a;
    while (cur != parents[cur]) {
      int next = parents[cur];
      parents[cur] = root;
      cur = parents[cur];
    }
    return root;
  }

  public void union(int a, int b) {
    int root_a = find(a);
    int root_b = find(b);
    if (root_a == root_b) return;
    int bigger = sizes[root_a] >= sizes[root_b] ? root_a : root_b;
    int smaller = sizes[root_a] < sizes[root_b] ? root_a : root_b;
    sizes[bigger] += sizes[smaller];
    parents[smaller] = parents[bigger];
  }
}

File: python/graphs/UnionFind.py
class UnionFind:
    def __init__(self, size: int):
        self.parents = [i for i in range(size)]
        self.sizes = [1 for i in range(size)]

    def union(self, v: int, u: int):
        v = find(v)
        u = find(u)
        if find(v) == find(u): 
            return

        if self.sizes[v] > self.size[u]:
            self.parents[u] = self.parents[v]
            self.sizes[u] += self.sizes[v]
            self.sizes[v] = self.sizes[u]
        else:
            self.parents[v] = self.parents[u]
            self.sizes[v] += self.sizes[u]
            self.sizes[u] = self.sizes[v]

    def find(self, v: int):
        cur = v
        while cur != self.parents[cur]:
            cur = self.parents[cur]

        root = cur
        # Path compression
        cur = v
        while cur != self.parents[cur]:
            par = self.parents[cur]
            self.parents[cur] = root
            cur = par

        return root
